### DISCUSSION TITLE
Run middleware only for `process.server`

### DISCUSSION THREAD

**Discussion by adriaandotcom** [nuxt/nuxt#17534] [Questions] [answered]
I'm trying to get a country code (`US`) in the client. The IP is only accessible from the server process.

I got this far:

```js
import { Reader } from "@maxmind/geoip2-node";
import { readFileSync } from "fs";

let dbBuffer;
if (process.server) {
  dbBuffer = readFileSync("./geolite-country.mmdb");
}

export default defineNuxtRouteMiddleware(() => {
  if (!process.server) return;

  const app = useNuxtApp();
  const { req: request } = app.ssrContext;
  const ip =
    request?.connection?.remoteAddress ||
    request?.connection?.socket?.remoteAddress;

  if (ip && dbBuffer) {
    try {
      const reader = Reader.openBuffer(dbBuffer);
      const response = reader.country(ip);
      console.log(response.country.isoCode);
    } catch (error) {
      console.error(error);
    }
  }
});
```

But the front end I get:

<img width="475" alt="image" src="https://user-images.githubusercontent.com/1079135/169833824-13d6da07-2629-4122-bf14-79378042510c.png">

**Comment by danielroe** [comment, marked as answer]
This should likely be a plugin rather than a middleware, and you can use `~/plugins/ip.server.ts` to load it only on server side.

**Comment by pi0** [comment]
In addition to the suggestion to use `plugins/ip.server` (which I consider a right answer), you might also use dynamic imports or a [Nitro Server Middleware](https://v3.nuxtjs.org/guide/features/server-routes#server-middleware) in `server/middleware` to do the checks or add country to `event.context.country` for future use.

Regardless, please try and consider, fs paths especially relative paths like `./geolite-country.mmdb` are probably not working with the production build which we might come with new a solution (my old solution with [maxmind-databases](https://github.com/pi0/maxmind-databases) is not working anymore due to license changes)

**Reply by adriaandotcom** [reply]
Thanks! I created a separate question for the local file: https://github.com/nuxt/framework/discussions/5174

**Comment by adriaandotcom** [comment]
Thanks both! This seems to work perfectly:

```js
// plugins/ip.server.js

import { Reader } from "@maxmind/geoip2-node";
import { readFileSync } from "fs";

const dbBuffer = readFileSync("./geolite-country.mmdb");

export default defineNuxtPlugin((nuxtApp) => {
  const { req: request } = nuxtApp.ssrContext;
  const ip =
    request?.connection?.remoteAddress ||
    request?.connection?.socket?.remoteAddress;

  if (ip && dbBuffer) {
    try {
      const reader = Reader.openBuffer(dbBuffer);
      const response = reader.country(ip);
      console.log({ found: true, code: response.country.isoCode });
    } catch (error) {
      console.log({ found: false, error: error.message });
    }
  } else {
    console.log({ found: false });
  }
});
```
